src/consensus/pbft/pbft.h

```diff
@@ -339,13 +345,36 @@ namespace pbft
       ITimer::handle_timeouts(elapsed);
     }
 
+    template <typename T>
+    size_t replicate_to_ledger(const T& data)
+    {
+      ledger->put_entry(data->data(), data->size());
+      return data->size();
+    }
+
+    template <>
+    size_t replicate_to_ledger<std::vector<uint8_t>>(
+      const std::vector<uint8_t>& data)
+    {
+      ledger->put_entry(data);
+      return data.size();
+    }
+
     bool replicate(const kv::BatchDetachedBuffer& entries) override
     {
+      for (auto&& [index, data, globally_committable] : entries)
+      {
+        replicate_to_ledger(data);
```

Comment:
Minor: should this be caller `store_in_ledger()` or similar as this does not actually replicate (as opposed to raft's replicate).